# Borrar columnas

Usando dataset ../Data/01CrudoNoEditar/01desastres_crudo_renamedcols.csv

Es copia de 01desastres_crudo.csv pero habiendo renombrado antes las cols con ('000 US$) directamente desde el csv, ya que .rename() da error

Objetivos:

Crear dataset 02desastres_colsreducidas.csv en ../Data/02ParaLimpiar/ sin columnas innecesarias

## Importar librerías

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Cargar datos

In [2]:
df = pd.read_csv('../../Data/01CrudoNoEditar/01desastres_crudo_renamedcols.csv', encoding='utf-8', delimiter=';', engine='python')
# encoding='utf8' da problemas
# dejar engine='python' para evitar advertencias

# Estadísticas básicas (columnas originales)

### Muestras head

In [3]:
df.head(5)

Dis No  Year   Seq Glide Disaster Group Disaster Subgroup  \
0  1900-9002-CPV  1900  9002   NaN        Natural    Climatological   
1  1900-9001-IND  1900  9001   NaN        Natural    Climatological   
2  1902-0012-GTM  1902    12   NaN        Natural       Geophysical   
3  1902-0003-GTM  1902     3   NaN        Natural       Geophysical   
4  1902-0010-GTM  1902    10   NaN        Natural       Geophysical   

       Disaster Type Disaster Subtype Disaster Subsubtype   Event Name  ...  \
0            Drought          Drought                 NaN          NaN  ...   
1            Drought          Drought                 NaN          NaN  ...   
2         Earthquake  Ground movement                 NaN          NaN  ...   
3  Volcanic activity         Ash fall                 NaN  Santa Maria  ...   
4  Volcanic activity         Ash fall                 NaN  Santa Maria  ...   

  Reconstruction Costs Adj Insured Damages Insured Damages Adj Total Damages  \
0                      NaN             NaN                 NaN           NaN   
1                      NaN             NaN                 NaN           NaN   
2                      NaN             NaN                 NaN       25000.0   
3                      NaN             NaN                 NaN           NaN   
4                      NaN             NaN                 NaN           NaN   

  Total Damages Adj          CPI Adm Level Admin1 Code Admin2 Code  \
0               NaN  2,849084409       NaN         NaN         NaN   
1               NaN  2,849084409       NaN         NaN         NaN   
2          843726.0  2,963047785       NaN         NaN         NaN   
3               NaN  2,963047785       NaN         NaN         NaN   
4               NaN  2,963047785       NaN         NaN         NaN   

  Geo Locations  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 50 columns]

## shape, info, dtypes

In [4]:
df.shape
# Resultado: 16636 filas y 50 columnas)

(16636, 50)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16636 entries, 0 to 16635
Data columns (total 50 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Dis No                    16636 non-null  object 
 1   Year                      16636 non-null  int64  
 2   Seq                       16636 non-null  int64  
 3   Glide                     1736 non-null   object 
 4   Disaster Group            16636 non-null  object 
 5   Disaster Subgroup         16636 non-null  object 
 6   Disaster Type             16636 non-null  object 
 7   Disaster Subtype          13313 non-null  object 
 8   Disaster Subsubtype       1117 non-null   object 
 9   Event Name                3969 non-null   object 
 10  Country                   16636 non-null  object 
 11  ISO                       16636 non-null  object 
 12  Region                    16636 non-null  object 
 13  Continent                 16636 non-null  object 
 14  Locati

In [6]:
df.dtypes

Dis No                       object
Year                          int64
Seq                           int64
Glide                        object
Disaster Group               object
Disaster Subgroup            object
Disaster Type                object
Disaster Subtype             object
Disaster Subsubtype          object
Event Name                   object
Country                      object
ISO                          object
Region                       object
Continent                    object
Location                     object
Origin                       object
Associated Dis               object
Associated Dis2              object
OFDA Response                object
Appeal                       object
Declaration                  object
AID Contribution            float64
Dis Mag Value               float64
Dis Mag Scale                object
Latitude                     object
Longitude                    object
Local Time                   object
River Basin                 

## Columnas originales

In [7]:
df.columns

Index(['Dis No', 'Year', 'Seq', 'Glide', 'Disaster Group', 'Disaster Subgroup',
       'Disaster Type', 'Disaster Subtype', 'Disaster Subsubtype',
       'Event Name', 'Country', 'ISO', 'Region', 'Continent', 'Location',
       'Origin', 'Associated Dis', 'Associated Dis2', 'OFDA Response',
       'Appeal', 'Declaration', 'AID Contribution', 'Dis Mag Value',
       'Dis Mag Scale', 'Latitude', 'Longitude', 'Local Time', 'River Basin',
       'Start Year', 'Start Month', 'Start Day', 'End Year', 'End Month',
       'End Day', 'Total Deaths', 'No Injured', 'No Affected', 'No Homeless',
       'Total Affected', 'Reconstruction Costs', 'Reconstruction Costs Adj',
       'Insured Damages', 'Insured Damages Adj', 'Total Damages',
       'Total Damages Adj', 'CPI', 'Adm Level', 'Admin1 Code', 'Admin2 Code',
       'Geo Locations'],
      dtype='object')

### Columnas con valores numéricos: 21

Year                                           int64, 
Seq                                            int64, 
AID Contribution                             float64, 
Dis Mag Value                                float64, 
Year                                         int64, 
Start Month                                  float64, 
Start Day                                    float64, 
End Year                                       int64, 
End Month                                    float64, 
End Day                                      float64, 
Total Deaths                                 float64, 
No Injured                                   float64, 
No Affected                                  float64, 
No Homeless                                  float64, 
Total Affected                               float64, 
Reconstruction Costs                         float64, 
Reconstruction Costs Adj                     float64, 
Insured Damages                              float64, 
Insured Damages Adj                          float64, 
Total Damages                                float64, 
Total Damages Adj                            float64

### Columnas con variables categóricas: 28
 
 3   Glide                     1736 non-null   object 
 4   Disaster Group            16636 non-null  object 
 5   Disaster Subgroup         16636 non-null  object 
 6   Disaster Type             16636 non-null  object 
 7   Disaster Subtype          13313 non-null  object 
 8   Disaster Subsubtype       1117 non-null   object 
 9   Event Name                3969 non-null   object 
 10  Country                   16636 non-null  object 
 11  ISO                       16636 non-null  object 
 12  Region                    16636 non-null  object 
 13  Continent                 16636 non-null  object 
 14  Location                  14825 non-null  object 
 15  Origin                    4085 non-null   object 
 16  Associated Dis            3593 non-null   object 
 17  Associated Dis2           763 non-null    object 
 18  OFDA Response             1716 non-null   object 
 19  Appeal                    2559 non-null   object 
 20  Declaration               3343 non-null   object
 23  Dis Mag Scale             15416 non-null  object 
 24  Latitude                  2775 non-null   object 
 25  Longitude                 2775 non-null   object 
 26  Local Time                1156 non-null   object 
 27  River Basin               1336 non-null   object
 45  CPI                       16530 non-null  object 
 46  Adm Level                 8475 non-null   object 
 47  Admin1 Code               5030 non-null   object 
 48  Admin2 Code               4248 non-null   object 
 49  Geo Locations             8475 non-null   object 

## Categorización de variables numéricas y categóricas por niveles según clasificación de dataset

In [41]:
num_lvl1 = ['Year', 'Seq']
categ_lvl1_id = ['Dis No']
categ_lvl1 = ['Disaster Subgroup', 'Disaster Type','Disaster Subtype']
categ_lvl2_geo = ['Country', 'ISO', 'Region', 'Continent', 'Location'] # 'River Basin'
categ_lvl2_geoloc = ['Latitude', 'Longitude'] # Estas pertenecen a la anterior pero es mejor separarlas debido a su carácter más específico
num_lvl2_temp = ['Start Year','Start Month', 'End Year', 'End Month'] # Son objects, evaluar si es necesario convertir a datetimes numéricas
# De num_lvl2_temp eliminar: 'Local Time', 'Start Day', 'End Day'
# categ_lvl2_time = ['Local Time']
categ_lvl2_fis = ['Origin','Dis Mag Scale']  # Se borrarán 'Associated Dis','Associated Dis2'
# categ_lvl2_status = ['AID Contribution', 'OFDA Response', 'Appeal','Declaration'] # Coincidir con World Risk Index
# categ_lvl3_orgs = [] Borrar Adm Level, Admin1 Code y Admin2 Code
# Las siguientes de lvl3 human_impact y economic_impact se sacarían de dataset complementario:
# num_lvl3_human_impact = ['Total Deaths', 'No Injured', 'No Affected', 'No Homeless', 'Total Affected']
# num_lvl3_economic_impact = ['Reconstruction Costs', 'Reconstruction Costs Adj', 'Insured Damages', 'Insured Damages Adj', 'Total Damages', 'Total Damages Adj']
num_lvl3_impact = ['Total Deaths', 'Total Affected', 'Total Damages', 'Total Damages Adj'] # Se dejan solo las que no tienen tantos nulls
# Faltan columnas (aquí se han puesto 43)

In [8]:
df = df.astype({"Start Day":"int","Start Month":"int", "End Day":"int","End Month":"int"})

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

## Criterio de borrado de columnas:
* Demasiados nulls
* Categoría de columnas más completa en dataset complementario
* Reducir dataset para acotar proyecto

## Valores faltantes

In [10]:
df.isnull().sum()

Dis No                          0
Year                            0
Seq                             0
Glide                       14900
Disaster Group                  0
Disaster Subgroup               0
Disaster Type                   0
Disaster Subtype             3323
Disaster Subsubtype         15519
Event Name                  12667
Country                         0
ISO                             0
Region                          0
Continent                       0
Location                     1811
Origin                      12551
Associated Dis              13043
Associated Dis2             15873
OFDA Response               14920
Appeal                      14077
Declaration                 13293
AID Contribution            15860
Dis Mag Value               11572
Dis Mag Scale                1220
Latitude                    13861
Longitude                   13861
Local Time                  15480
River Basin                 15300
Start Year                      0
Start Month   

In [9]:
df.isnull()

Dis No   Year    Seq  Glide  Disaster Group  Disaster Subgroup  \
0       False  False  False   True           False              False   
1       False  False  False   True           False              False   
2       False  False  False   True           False              False   
3       False  False  False   True           False              False   
4       False  False  False   True           False              False   
...       ...    ...    ...    ...             ...                ...   
16631   False  False  False   True           False              False   
16632   False  False  False  False           False              False   
16633   False  False  False   True           False              False   
16634   False  False  False   True           False              False   
16635   False  False  False  False           False              False   

       Disaster Type  Disaster Subtype  Disaster Subsubtype  Event Name  ...  \
0              False             False                 True        True  ...   
1              False             False                 True        True  ...   
2              False             False                 True        True  ...   
3              False             False                 True       False  ...   
4              False             False                 True       False  ...   
...              ...               ...                  ...         ...  ...   
16631          False             False                 True        True  ...   
16632          False             False                 True       False  ...   
16633          False             False                 True       False  ...   
16634          False             False                 True        True  ...   
16635          False             False                 True       False  ...   

       Reconstruction Costs Adj  Insured Damages  Insured Damages Adj  \
0                          True             True                 True   
1                          True             True                 True   
2                          True             True                 True   
3                          True             True                 True   
4                          True             True                 True   
...                         ...              ...                  ...   
16631                      True             True                 True   
16632                      True             True                 True   
16633                      True             True                 True   
16634                      True             True                 True   
16635                      True             True                 True   

       Total Damages  Total Damages Adj    CPI  Adm Level  Admin1 Code  \
0               True               True  False       True         True   
1               True               True  False       True         True   
2              False              False  False       True         True   
3               True               True  False       True         True   
4               True               True  False       True         True   
...              ...                ...    ...        ...          ...   
16631           True               True   True      False        False   
16632           True               True   True      False        False   
16633           True               True   True       True         True   
16634           True               True   True       True         True   
16635           True               True   True      False        False   

       Admin2 Code  Geo Locations  
0             True           True  
1             True           True  
2             True           True  
3             True           True  
4             True           True  
...            ...            ...  
16631        False          False  
16632         True          False  
16633         True           True  
16634         True           True  
16635  

### Columnas (numéricas y categóricas) con más valores faltantes

Numéricas con más nulls:
AID Contribution            15860
Dis Mag Value               11572
No Injured                  12489
No Homeless                 14166
Reconstruction Costs        16598
Reconstruction Costs Adj    16600
Insured Damages             15527
Insured Damages Adj         15527
Total Damages               11252
Total Damages Adj           11270

Categóricas con más nulls:
Origin                      12551
Associated Dis              13043
Associated Dis2             15873
OFDA Response               14920
Appeal                      14077
Declaration                 13293
Latitude                    13861
Longitude                   13861
Local Time                  15480
River Basin                 15300

## Borrar columnas

In [11]:
df.drop(columns=['Event Name', 'Disaster Group', 'Disaster Subsubtype', 'Glide', 'Local Time', 'River Basin', 'No Injured', 'No Affected', 'No Homeless', 'Reconstruction Costs', 'Reconstruction Costs Adj', 'Insured Damages', 'Insured Damages Adj', 'Adm Level', 'Admin1 Code', 'Admin2 Code', 'Geo Locations', 'Associated Dis', 'Associated Dis2', 'OFDA Response', 'Appeal', 'Declaration', 'AID Contribution', 'Year', 'CPI'], inplace=True)

In [12]:
print(df)

              Dis No   Seq Disaster Subgroup      Disaster Type  \
0      1900-9002-CPV  9002    Climatological            Drought   
1      1900-9001-IND  9001    Climatological            Drought   
2      1902-0012-GTM    12       Geophysical         Earthquake   
3      1902-0003-GTM     3       Geophysical  Volcanic activity   
4      1902-0010-GTM    10       Geophysical  Volcanic activity   
...              ...   ...               ...                ...   
16631  2023-0110-ZMB   110      Hydrological              Flood   
16632  2023-0068-ZMB    68        Biological           Epidemic   
16633  2023-0095-ZWE    95    Meteorological              Storm   
16634  2023-0022-SRB    22      Hydrological              Flood   
16635  2023-0140-SSD   140        Biological           Epidemic   

        Disaster Subtype      Country  ISO           Region Continent  \
0                Drought   Cabo Verde  CPV   Western Africa    Africa   
1                Drought        India  IND    Sou

## Estadísticas básicas posteriores a borrar columnas

In [13]:
df.shape #(16636, 25)

(16636, 25)

In [14]:
df.columns
'''
Index(['Dis No', 'Seq', 'Disaster Subgroup', 'Disaster Type', 'Disaster Subtype', 'Country', 'ISO', 'Region', 'Continent', 'Location', 'Origin', 'Dis Mag Value', 'Dis Mag Scale', 'Latitude', 'Longitude', 'Start Year', 'Start Month', 'End Year', 'End Month', 'CPI'], dtype='object')
'''

"\nIndex(['Dis No', 'Seq', 'Disaster Subgroup', 'Disaster Type',\n       'Disaster Subtype', 'Country', 'ISO', 'Region', 'Continent', 'Location',\n       'Origin', 'Dis Mag Value', 'Dis Mag Scale', 'Latitude', 'Longitude',\n       'Start Year', 'Start Month', 'End Year', 'End Month', 'CPI'],\n      dtype='object')\n"

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16636 entries, 0 to 16635
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Dis No             16636 non-null  object 
 1   Seq                16636 non-null  int64  
 2   Disaster Subgroup  16636 non-null  object 
 3   Disaster Type      16636 non-null  object 
 4   Disaster Subtype   13313 non-null  object 
 5   Country            16636 non-null  object 
 6   ISO                16636 non-null  object 
 7   Region             16636 non-null  object 
 8   Continent          16636 non-null  object 
 9   Location           14825 non-null  object 
 10  Origin             4085 non-null   object 
 11  Dis Mag Value      5064 non-null   float64
 12  Dis Mag Scale      15416 non-null  object 
 13  Latitude           2775 non-null   object 
 14  Longitude          2775 non-null   object 
 15  Start Year         16636 non-null  int64  
 16  Start Month        162

In [16]:
df.head()

Dis No   Seq Disaster Subgroup      Disaster Type Disaster Subtype  \
0  1900-9002-CPV  9002    Climatological            Drought          Drought   
1  1900-9001-IND  9001    Climatological            Drought          Drought   
2  1902-0012-GTM    12       Geophysical         Earthquake  Ground movement   
3  1902-0003-GTM     3       Geophysical  Volcanic activity         Ash fall   
4  1902-0010-GTM    10       Geophysical  Volcanic activity         Ash fall   

      Country  ISO           Region Continent                   Location  ...  \
0  Cabo Verde  CPV   Western Africa    Africa                Countrywide  ...   
1       India  IND    Southern Asia      Asia                     Bengal  ...   
2   Guatemala  GTM  Central America  Americas  Quezaltenango, San Marcos  ...   
3   Guatemala  GTM  Central America  Americas                        NaN  ...   
4   Guatemala  GTM  Central America  Americas                        NaN  ...   

  Start Year  Start Month Start Day End Year End Month  End Day  Total Deaths  \
0       1900          NaN       NaN     1900       NaN      NaN       11000.0   
1       1900          NaN       NaN     1900       NaN      NaN     1250000.0   
2       1902          4.0      18.0     1902       4.0     18.0        2000.0   
3       1902          4.0       8.0     1902       4.0      8.0        1000.0   
4       1902         10.0      24.0     1902      10.0     24.0        6000.0   

   Total Affected  Total Damages  Total Damages Adj  
0             NaN            NaN                NaN  
1             NaN            NaN                NaN  
2             NaN        25000.0           843726.0  
3             NaN            NaN                NaN  
4             NaN            NaN                NaN  

[5 rows x 25 columns]

# Guardar dataset en csv

In [23]:
df.to_csv('../../Data/02ParaLimpiar/02desastres_colsreducidas.csv', index=False, sep=';', encoding='utf-8')

Dataset actualizado con reducción de columnas 26/06 (borrado de cols relacionadas con Status Administrativo, Year y Associated Dis). Falta limpiar nulls o cambiar por otros valores.